tests/phpunit/tests/bookmark/getBookmark.php

```diff
@@ -0,0 +1,384 @@
+
+		// Run the function and test results.
+		$actual_bookmark = get_bookmark( ...$args );
+
+		$this->assertSame( $expected, $actual_bookmark );
+
+		// Check the bookmark was cached.
+		$actual_cache = wp_cache_get( $bookmark->link_id, 'bookmark' );
+		$this->assertEquals( $bookmark, $actual_cache );
+	}
+
+	/**
+	 * Path 2 data provider, i.e. when bookmark instance is given.
+	 */
+	public function data_when_instance_bookmark() {
+		return array(
+			// Unhappy path.
+			'with incomplete bookmark data'      => array(
+				array(
+					'bookmark' => (object) array(
+						'link_id' => '100',
+					),
+				),
+			),
+			'with invalid output'                => array(
+				array(
+					'output' => 'invalid',
+				),
+			),
+			'with invalid filter'                => array(
+				array(
+					'filter' => 'invalid',
+				),
+			),
+			// Happy path.
+			'with defaults'                      => array(
+				array(),
+			),
+			'with non-default output'            => array(
+				array(
+					'output' => ARRAY_A,
+				),
+			),
+			'with non-default filter'            => array(
+				array(
+					'filter' => 'display',
+				),
+			),
+			'with non-default output and filter' => array(
+				array(
+					'output' => ARRAY_N,
+					'filter' => 'display',
+				),
+			),
+		);
+	}
+
+	/**
+	 * Path 3A: Uses the global link when exists and the given bookmark link ID matches the global link.
+	 *
+	 * @dataProvider data_when_else
+	 *
+	 * @param array $args Function argument list.
+	 */
+	public function test_should_return_global_when_else( $args ) {
+		$args            = $this->init_func_args( $args, $this->bookmark->link_id );
+		$GLOBALS['link'] = $this->bookmark;
+		$expected        = $this->maybe_format_expected_data( $args, $GLOBALS['link'] );
+
+		// Run the function and test results.
+		$actual_bookmark = get_bookmark( ...$args );
+
+		$this->assertSame( $expected, $actual_bookmark );
+		$this->assertFalse( wp_cache_get( $this->bookmark->link_id, 'bookmark' ) );
+	}
+
+	/**
+	 * Path 3B: Pulls from cache when given existing bookmark link ID.
+	 *
+	 * @dataProvider data_when_else
+	 *
+	 * @param array $args Function argument list.
+	 */
+	public function test_should_return_cached_bookmark_when_given_existing_link_id( $args ) {
+		// Cache the bookmark instance to setup the test.
+		wp_cache_add( $this->bookmark->link_id, $this->bookmark, 'bookmark' );
+		$args     = $this->init_func_args( $args, $this->bookmark->link_id );
+		$expected = $this->maybe_format_expected_data( $args, $this->bookmark );
+
+		// Run the function and test results.
+		$actual_bookmark = get_bookmark( ...$args );
+
+		// For non-array output type, use assetEquals. Why? The object pulled from cache will have the same
+		// property values but will be a different object than the expected object.
+		if ( is_object( $expected ) ) {
+			$this->assertEquals( $expected, $actual_bookmark );
+		} else {
+			$this->assertSame( $expected, $actual_bookmark );
```

Comment:
```suggestion
			$this->assertSameSets( $expected, $actual_bookmark );
```

Custom WP sniff for arrays in case the order chagnes. Maybe `assertSameSetsWithIndex()` if you prefer.

Reply:
Resolved in commit https://github.com/WordPress/wordpress-develop/pull/1180/commits/263f7cc3c7779ebfebafd33d731bb79623dd2382